javascript/test/text_v1.ts

```diff
@@ -294,4 +294,30 @@ describe("Automerge.Text", () => {
     s4 = Automerge.merge(s4, s2)
     assert.strictEqual(s4.text.toString(), "🐦")
   })
+
+  it("should let you insert strings", () => {
+    s1 = Automerge.from({
+      text: new Automerge.Text(""),
+    })
+
+    s1 = Automerge.change(s1, d => {
+      d.text.insertAt(0, "four")
+    })
+
+    assert.strictEqual(s1.text.length, 4)
+  })
+
+  it("should index by unicode code points", () => {
+    s1 = Automerge.from({
+      text: new Automerge.Text(""),
+    })
+
+    s1 = Automerge.change(s1, d => {
+      d.text.insertAt(0, "🇬🇧")
+      d.text.insertAt(2, "four")
```

Comment:
I have a question which may be my misunderstanding but I would have imagined the insertion here would be under index 4 as `"🇬🇧".length` is 4. 

What is the 2 here representing? If this was in fact insertion mid-codepoints, I would've expected `🇬four🇧` as the emoji `🇬🇧` is composed of those two regional indicators.

Can you please explain this test? mostly from the standpoint of a user of the library, as I have different expectations of what should happen here.

Reply:
It comes down to the way that Unicode handles flags, and the way that JavaScript handles Unicode.

The flag emoji is rendered as one glyph, but under the hood is represented by two unicode codepoints, (see https://emojipedia.org/flag-united-kingdom/), U+1F1EC and U+1F1E7.

automerge uses list positions based on codepoints and so 2 represents the position after the flag.

JavaScript was created early on in the history of Unicode when it was believed that there would be a maximum of 2**16 characters; and so it represented strings using a legacy encoding known as UCS-2 which uses two bytes per code point.

When Unicode realized that they needed more characters the UTF-16 encoding was created to be backward compatible with UCS-2. UTF-16 like UCS-2 uses two bytes to represent each “code unit”. For many characters there is an exact equivalence where one code unit represents the code point with the same value; but to represent more than 2**16 code points, some Unicode “code points” (including both the code points in the flag) are represented as a pair of two code units (see https://learn.microsoft.com/en-us/globalization/encoding/surrogate-pairs).

So the flag emoji is 4 code units long, and two code points long, but only conceptually one user perceivable glyph.

Reply:
I should also have mentioned that there is a plan to ship a new way to interact with automerge text that will let you use JavaScript string indices, but I’m not exactly sure when that will ship.

In the meantime, I’m using https://www.npmjs.com/package/code-point-mapping.